pkg/internal/ingress/ingress.go

```diff
@@ -0,0 +1,113 @@
+type InternalIngressInterface interface {
+	Create(ctx context.Context, ingress *networkingv1.Ingress, opts metav1.CreateOptions) (*networkingv1.Ingress, error)
+	Update(ctx context.Context, ingress *networkingv1.Ingress, opts metav1.UpdateOptions) (*networkingv1.Ingress, error)
+	Delete(ctx context.Context, name string, opts metav1.DeleteOptions) error
+}
+
+// InternalIngressLister mimics a client-go networking/v1/IngressLister.
+type InternalIngressLister interface {
+	List(selector labels.Selector) (ret []*networkingv1.Ingress, err error)
+	Ingresses(namespace string) InternalIngressNamespaceLister
+}
+
+// InternalIngressNamespaceLister mimics a client-go networking/v1/IngressNamespaceLister
+type InternalIngressNamespaceLister interface {
+	List(selector labels.Selector) (ret []*networkingv1.Ingress, err error)
+	Get(name string) (*networkingv1.Ingress, error)
+}
+
+// NewListerInformer returns an InternalIngressLister configured for v1 or v1beta1 ingresses depending on the
+// API Versions available in the discovery client.
+func NewListerInformer(ctx *controller.Context) (InternalIngressLister, cache.SharedIndexInformer, error) {
+	if hasVersion(ctx.Discovery, networkingv1.SchemeGroupVersion.String()) {
+		return &v1Lister{
+				lister: ctx.KubeSharedInformerFactory.Networking().V1().Ingresses().Lister(),
+			},
+			ctx.KubeSharedInformerFactory.Networking().V1().Ingresses().Informer(),
+			nil
+	} else if hasVersion(ctx.Discovery, networkingv1beta1.SchemeGroupVersion.String()) {
+		sch := runtime.NewScheme()
+		clientgoscheme.AddToScheme(sch)
+		return &v1beta1Lister{
+				scheme: sch,
+				lister: ctx.KubeSharedInformerFactory.Networking().V1beta1().Ingresses().Lister(),
+			},
+			ctx.KubeSharedInformerFactory.Networking().V1beta1().Ingresses().Informer(),
+			nil
+	}
+	return nil, nil, fmt.Errorf("neither %s or %s have any APIResources", networkingv1.SchemeGroupVersion, networkingv1beta1.SchemeGroupVersion)
+}
+
+// NewCreateUpdater returns an InternalIngressCreateUpdater configured for v1 or v1beta1 ingresses depending on the
+// versions available in the discovery client
+func NewCreateUpdater(ctx *controller.Context, discoveryOverrides ...discovery.DiscoveryInterface) (InternalIngressCreateUpdater, error) {
+	if hasVersion(ctx.Discovery, networkingv1.SchemeGroupVersion.String()) {
+		return &v1CreaterUpdater{
+			client: ctx.Client,
+		}, nil
+	} else if hasVersion(ctx.Discovery, networkingv1beta1.SchemeGroupVersion.String()) {
+		sch := runtime.NewScheme()
+		clientgoscheme.AddToScheme(sch)
+		return &v1beta1CreaterUpdater{
+			client: ctx.Client,
+			scheme: sch,
+		}, nil
+	} else {
+		return nil, fmt.Errorf("neither %s or %s have any APIResources", networkingv1.SchemeGroupVersion, networkingv1beta1.SchemeGroupVersion)
+	}
+}
+
+func hasVersion(d discovery.DiscoveryInterface, GroupVersion string) bool {
+	resourceList, err := d.ServerResourcesForGroupVersion(GroupVersion)
+	if err != nil {
+		return false
```

Comment:
This seems like it can cause issues when a transient network error occurs. 

In the event of an error, I would expect an `IsNotFound` or equivalent check, and to retry or error out completely otherwise.

Reply:
https://github.com/kubernetes/client-go/blob/release-1.21/discovery/discovery_client.go#L201-L206

This never returns an error on notfound or forbidden. What do you think would be the best thing to do? Try a few times and store first successful result in a cache?

Reply:
Understood, in that case we should error out completely as something has gone quite wrong..

Reply:
I just found out that `ServerResourcesForGroupVersion` returns a Forbidden error, which goes unnoticed since it is neither logged not bubbled up.

That happens quite often for me to use the wrong client certs, and I end up with 401s in cert-manager (I was playing with `mitmproxy` and `telepresence`).

I think `hasVersion` should return errors when the code is not 200 or 404:

```go
func hasVersion(d discovery.DiscoveryInterface, GroupVersion string) (bool, error) {
    // ...

    resourceList, err := d.ServerResourcesForGroupVersion(GroupVersion)
    if apierrors.IsNotFound(err) {
    	return false, nil
    }
    if err != nil {
    	return false, err
    }

    // ...
}
```